language_service/src/hover.rs

```diff
@@ -204,26 +300,68 @@ impl Visitor<'_> for HoverVisitor<'_> {
     }
 }
 
-fn markdown_with_doc(doc: &str, namespace: Option<Rc<str>>, code: impl Display) -> String {
-    let parsed_doc = parse_doc(doc);
+fn protocol_span(span: qsc::Span) -> protocol::Span {
+    protocol::Span {
+        start: span.lo,
+        end: span.hi,
+    }
+}
+
+fn curr_callable_to_params(curr_callable: Option<&ast::CallableDecl>) -> Vec<&ast::Pat> {
+    match curr_callable {
+        Some(decl) => vec![decl.input.as_ref()],
+        None => vec![],
+    }
+}
 
-    let code = match namespace {
-        Some(namespace) if !namespace.is_empty() => {
-            markdown_fenced_block(format!("{namespace}\n{code}"))
+fn is_param(param_pats: &[&ast::Pat], node_id: ast::NodeId) -> bool {
+    fn find_in_pat(pat: &ast::Pat, node_id: ast::NodeId) -> bool {
+        match &*pat.kind {
+            ast::PatKind::Bind(ident, _) => node_id == ident.id,
+            ast::PatKind::Discard(_) | ast::PatKind::Elided => false,
+            ast::PatKind::Paren(inner) => find_in_pat(inner, node_id),
+            ast::PatKind::Tuple(inner) => inner.iter().any(|x| find_in_pat(x, node_id)),
         }
-        _ => markdown_fenced_block(code),
-    };
+    }
 
-    if parsed_doc.summary.is_empty() {
-        code
-    } else {
-        format!("{}---\n{}\n", code, parsed_doc.summary,)
+    param_pats.iter().any(|pat| find_in_pat(pat, node_id))
+}
+
+fn display_local(
+    param_kind: &LocalKind,
+    markdown: &String,
+    local_name: &str,
+    callable_name: &str,
+    callable_doc: &str,
+) -> String {
+    match param_kind {
+        LocalKind::Param => {
+            let param_doc = parse_doc_for_param(callable_doc, local_name);
+            with_doc(
+                &param_doc,
+                format!("param of `{callable_name}`\n{markdown}",),
+            )
+        }
+        LocalKind::LambdaParam => format!("lambda param\n{markdown}"),
+        LocalKind::Local => format!("local\n{markdown}"),
```

Comment:
Ehh, I think we can skip the "local" descriptor for a more uncluttered look. I don't think it adds any new information for the user.

Reply:
My end-goal for these is to have the hover for each thing tell the user the _kind_ of object they are hovering over in a consistent way, i.e. `parameter`, `local`, `operation`, `function`, `newtype`, `field`, and any additional kinds. These would appear left of the name in blue (as is currently the case for `operation`, `function`, and `newtype`), with the container name above (namespace/module name, callable name, UDT name, "lambda"). It would be even better if we could add links to the hover card for people to click through to definition for those callables and UDTs.

Reply:
I'm not currently sure how to make the text appear blue in the same way as the `operation` or `function` though, as that is done by the fact that the markdown is using qsharp keyword recognition.

Reply:
I don't think there's an easy way for us to provide formatting besides Q# syntax highlighting. Taking a page from TypeScript, instead of adding a word like "local", they synthesize some code like `let foo = 3;` to show how the local was declared. ([See this example ](https://www.typescriptlang.org/play?#code/PQKhCgAIUgBAHAhgJ0QW0mgnk1GAKK6ApgC7HKQAmA9gMYCuaxAdqYqQJY0tQjDgAZgxZ0uPTFmGjxLABTZc6AFyQAyqWScWAcwCUkAN5RIpujwDOpSeZZXIAXkgAiABadnAbhOmANmUkANxRHF1diX18aLx9IW3tfdAAjKkRQuUS0FMQlNGUNLV1IAwcAPiNITOzcyABfb1jFIjRvU1NseNJWtuxg5G9a8HBsaTFuFiA) - hover over the names). I checked Rust and they basically do the same thing. That looks similar to how we format `operation` and `function`s - we make up some hover text that looks just like declaration itself. What do you think of doing this for locals too?

For parameters, I see TypeScript just tacks on `(parameter)` and Rust doesn't bother calling them out. But I like your solution the best of all three.

Reply:
Discussed design offline. First line of hover card will be location information for the symbol, i.e. "local", "parameter of ...", namespace, etc. Second line will be signature. Subsequent lines will be doc comments.